rust/src/tokio/read.rs

```diff
@@ -0,0 +1,455 @@
+        match self.get_mut() {
+            ReaderState::Base(r) => pin!(r).poll_read(cx, buf),
+            ReaderState::UncompressedChunk(r) => pin!(r).poll_read(cx, buf),
+            ReaderState::ZstdChunk(r) => pin!(r).poll_read(cx, buf),
+            ReaderState::Lz4Chunk(r) => pin!(r).poll_read(cx, buf),
+            ReaderState::Empty => {
+                panic!("invariant: reader is only set to empty while swapping with another valid variant")
+            }
+        }
+    }
+}
+impl<R> ReaderState<R>
+where
+    R: AsyncRead,
+{
+    pub fn into_inner(self) -> McapResult<R> {
+        match self {
+            ReaderState::Base(reader) => Ok(reader),
+            ReaderState::UncompressedChunk(take) => Ok(take.into_inner()),
+            ReaderState::ZstdChunk(decoder) => Ok(decoder.into_inner().into_inner().into_inner()),
+            ReaderState::Lz4Chunk(decoder) => {
+                let (output, result) = decoder.finish();
+                result?;
+                Ok(output.into_inner())
+            }
+            ReaderState::Empty => {
+                panic!("invariant: reader is only set to empty while swapping with another valid variant")
+            }
+        }
+    }
+}
+/// Reads an MCAP file record-by-record, writing the raw record data into a caller-provided Vec.
+pub struct RecordReader<R> {
+    reader: ReaderState<R>,
+    options: Options,
+    start_magic_seen: bool,
+    footer_seen: bool,
+    to_discard_after_chunk: usize,
+    scratch: Box<[u8]>,
+}
+
+#[derive(Default, Clone)]
+pub struct Options {
+    /// If true, the reader will not expect the MCAP magic at the start of the stream.
+    skip_start_magic: bool,
+    /// If true, the reader will not expect the MCAP magic at the end of the stream.
+    skip_end_magic: bool,
+    // If true, the reader will yield entire chunk records. Otherwise, the reader will decompress
+    // and read into the chunk, yielding the records inside.
+    emit_chunks: bool,
+}
+
+enum Cmd {
+    YieldRecord(u8),
+    EnterChunk {
+        header: records::ChunkHeader,
+        len: u64,
+    },
+    ExitChunk,
+    Stop,
+}
+
+impl<R> RecordReader<R>
+where
+    R: AsyncRead + std::marker::Unpin,
+{
+    pub fn new(reader: R) -> Self {
+        Self::new_with_options(reader, &Options::default())
+    }
+
+    pub fn new_with_options(reader: R, options: &Options) -> Self {
+        Self {
+            reader: ReaderState::Base(reader),
+            options: options.clone(),
+            start_magic_seen: false,
+            footer_seen: false,
+            to_discard_after_chunk: 0,
+            scratch: vec![0; 1024].into_boxed_slice(),
+        }
+    }
+
+    pub fn into_inner(self) -> McapResult<R> {
+        self.reader.into_inner()
+    }
+
+    /// Reads the next record from the input stream and copies the raw content into `data`.
+    /// Returns the record's opcode as a result.
+    pub async fn next_record(&mut self, data: &mut Vec<u8>) -> McapResult<Option<u8>> {
```

Comment:
this is not a stream, because the user needs to be able to pass their own backing buffer in per record.

For discussion:
* streams yield Option<result<>> rather than Result<Option<>>, should I do that instead?
* I could implement this as a stream that returns (opcode, Vec) per item, and if the user wants to reuse a vec between calls they can pass it in with another function, ie:
```rust

while let Some((opcode, data)) = record_reader.next_record().await? {
   /// ... do something with the data
   record_reader.set_next_data_buf(data);
}
```